File: backend/app/parsers/novelai.py
import contextlib
import json
from typing import Any

from app.parsers.base import (
    MetadataParser,
    ModelType,
    ParsedMetadata,
    SourceTool,
)


class NovelAIParser(MetadataParser):
    """Parser for NovelAI metadata format."""

    def can_parse(self, png_info: dict[str, Any]) -> bool:
        """Check if the PNG info contains NovelAI metadata."""
        if "Comment" not in png_info:
            return False

        try:
            comment = png_info["Comment"]
            data = json.loads(comment) if isinstance(comment, str) else comment

            # Check for NovelAI-specific keys
            return isinstance(data, dict) and ("uc" in data or "prompt" in data)
        except (json.JSONDecodeError, TypeError):
            return False

    def parse(self, png_info: dict[str, Any]) -> ParsedMetadata:
        """Parse NovelAI metadata."""
        comment_str = png_info.get("Comment", "{}")
        data = json.loads(comment_str) if isinstance(comment_str, str) else comment_str

        metadata = ParsedMetadata(
            source_tool=SourceTool.NOVELAI,
            has_metadata=True,
            raw_metadata=data,
            model_type=ModelType.OTHER,  # NovelAI uses custom models
        )

        # Extract prompts
        if "prompt" in data:
            metadata.positive_prompt = str(data["prompt"])

        if "uc" in data:
            metadata.negative_prompt = str(data["uc"])

        # Extract parameters
        if "steps" in data:
            with contextlib.suppress(ValueError, TypeError):
                metadata.steps = int(data["steps"])

        if "scale" in data:
            with contextlib.suppress(ValueError, TypeError):
                metadata.cfg_scale = float(data["scale"])

        if "seed" in data:
            with contextlib.suppress(ValueError, TypeError):
                metadata.seed = int(data["seed"])

        if "sampler" in data:
            # NovelAI uses k_ prefix for samplers
            sampler = str(data["sampler"])
            if sampler.startswith("k_"):
                sampler = sampler[2:]
            metadata.sampler_name = sampler

        # Store additional NovelAI-specific data in model_params
        model_params: dict[str, Any] = {}

        for key in ["width", "height", "n_samples", "ucPreset", "qualityToggle"]:
            if key in data:
                model_params[key] = data[key]

        if model_params:
            metadata.model_params = model_params

        return metadata

File: backend-rs/src/parser/novelai.rs
//! NovelAI parser (mirror parsers/novelai.py). Metadata lives in the PNG
//! `Comment` text chunk as a JSON object.

use std::collections::HashMap;

use serde_json::{json, Value};

use super::{ModelType, ParsedMetadata, SourceTool};

pub fn can_parse(png_info: &HashMap<String, String>) -> bool {
    let Some(comment) = png_info.get("Comment") else {
        return false;
    };
    match serde_json::from_str::<Value>(comment) {
        Ok(data) => data.is_object() && (data.get("uc").is_some() || data.get("prompt").is_some()),
        Err(_) => false,
    }
}

pub fn parse(png_info: &HashMap<String, String>) -> ParsedMetadata {
    let comment = png_info.get("Comment").map(String::as_str).unwrap_or("{}");
    let data: Value = serde_json::from_str(comment).unwrap_or_else(|_| json!({}));

    let mut m = ParsedMetadata::new(SourceTool::Novelai);
    m.model_type = Some(ModelType::Other);
    m.raw_metadata = Some(data.clone());

    if let Some(p) = data.get("prompt") {
        m.positive_prompt = Some(value_to_string(p));
    }
    if let Some(uc) = data.get("uc") {
        m.negative_prompt = Some(value_to_string(uc));
    }
    if let Some(v) = data.get("steps").and_then(as_loose_i64) {
        m.steps = Some(v as i32);
    }
    if let Some(v) = data.get("scale").and_then(as_loose_f64) {
        m.cfg_scale = Some(v);
    }
    if let Some(v) = data.get("seed").and_then(as_loose_i64) {
        m.seed = Some(v);
    }
    if let Some(sampler) = data.get("sampler") {
        // NovelAI prefixes sampler names with "k_".
        let s = value_to_string(sampler);
        m.sampler_name = Some(s.strip_prefix("k_").unwrap_or(&s).to_string());
    }

    for key in ["width", "height", "n_samples", "ucPreset", "qualityToggle"] {
        if let Some(v) = data.get(key) {
            m.model_params.insert(key.to_string(), v.clone());
        }
    }

    m
}

/// Mirror Python `str(value)` for prompt/sampler fields: strings pass through,
/// everything else uses its JSON representation.
fn value_to_string(v: &Value) -> String {
    match v {
        Value::String(s) => s.clone(),
        other => other.to_string(),
    }
}

fn as_loose_i64(v: &Value) -> Option<i64> {
    v.as_i64()
        .or_else(|| v.as_f64().map(|f| f as i64))
        .or_else(|| v.as_str().and_then(|s| s.parse().ok()))
}

fn as_loose_f64(v: &Value) -> Option<f64> {
    v.as_f64().or_else(|| v.as_str().and_then(|s| s.parse().ok()))
}
